src/Libraries/Microsoft.Extensions.DataIngestion/Chunkers/ElementsChunker.cs

```diff
@@ -0,0 +1,266 @@
+
+                    // some tokens fit
+                    if (index > 0)
+                    {
+                        // We could try to split by sentences or other delimiters, but it's complicated.
+                        // For simplicity, we will just split at the last new line that fits.
+                        // Our promise is not to go over the max token count, not to create perfect chunks.
+                        int newLineIndex = remainingContent.Slice(0, index).LastIndexOf('\n');
+                        if (newLineIndex > 0)
+                        {
+                            index = newLineIndex + 1; // We want to include the new line character (works for "\r\n" as well).
+                            tokenCount = CountTokens(remainingContent.Slice(0, index));
+                        }
+
+                        totalTokenCount += tokenCount;
+                        ReadOnlySpan<char> spanToAppend = remainingContent.Slice(0, index);
+                        AppendNewLineAndSpan(_currentChunk, spanToAppend);
+                        remainingContent = remainingContent.Slice(index);
+                    }
+                    else if (totalTokenCount == contextTokenCount)
+                    {
+                        // We are at the beginning of a chunk, and even a single token does not fit.
+                        ThrowTokenCountExceeded();
+                    }
+
+                    if (!remainingContent.IsEmpty)
+                    {
+                        Commit();
+                    }
+                }
+            }
+
+            if (totalTokenCount == _maxTokensPerChunk)
+            {
+                Commit();
+            }
+        }
+
+        if (totalTokenCount > contextTokenCount)
+        {
+            chunks.Add(new(_currentChunk.ToString(), document, context));
+        }
+
+        _currentChunk = _currentChunk.Clear();
+
+        return chunks;
+
+        void Commit()
+        {
+            chunks.Add(new(_currentChunk.ToString(), document, context));
+
+            // We keep the context in the current chunk as it's the same for all elements.
+            _currentChunk = _currentChunk.Remove(
+                startIndex: context.Length,
+                length: _currentChunk.Length - context.Length);
+            totalTokenCount = contextTokenCount;
+        }
+
+        static void ThrowTokenCountExceeded()
+            => throw new InvalidOperationException("Can't fit in the current chunk. Consider increasing max tokens per chunk.");
+    }
+
+    private static void AppendNewLineAndSpan(StringBuilder stringBuilder, ReadOnlySpan<char> chars)
+    {
+        // Don't start an empty chunk (no context provided) with a new line.
+        if (stringBuilder.Length > 0)
+        {
+            stringBuilder.AppendLine();
+        }
+
+#if NET
+        stringBuilder.Append(chars);
+#else
+        stringBuilder.Append(chars.ToString());
```

Comment:
You could do:
```csharp
unsafe
{
    fixed (char* charsPtr = chars)
    {
        stringBuilder.Append(charsPtr, chars.Length);
    }
}
```
to avoid materializing the extra string.